sycl/doc/extensions/proposed/sycl_ext_oneapi_get_kernel_info.asciidoc

```diff
@@ -0,0 +1,228 @@
+----
+!====
+
+_Constraints:_ Available only if `Param` is an information descriptor for the
+`kernel` class, which can be used by the `kernel::get_info(const device &)`
+overload.
+
+_Preconditions:_ The `KernelName` must be the type kernel name of a kernel that
+is defined in the application.
+The device `dev` must be one of the devices contained by `ctxt`.
+The kernel `KernelName` must be compatible with the device `dev` as defined by
+`is_compatible`.
+
+_Returns:_ The same value `ret` that would be computed by:
+
+[source,c++]
+----
+auto bundle =
+  sycl::get_kernel_bundle<KernelName, sycl::bundle_state::executable>(ctxt);
+auto ret = bundle.get_kernel<KernelName>().get_info<Param>(dev);
+----
+
+_Remarks:_ Each information descriptor may specify additional preconditions,
+exceptions that are thrown, etc.
+
+'''
+
+[frame=all,grid=none,separator="@"]
+!====
+a@
+[source,c++]
+----
+namespace sycl::ext::oneapi {
+
+template<typename KernelName, typename Param>
+typename Param::return_type get_kernel_info(const queue& q);
+
+} // namespace sycl::ext::oneapi
+----
+!====
+
+_Constraints:_ Available only if `Param` is an information descriptor for the
+`kernel` class, which can be used by the `kernel::get_info(const device &)`
+overload.
+
+_Preconditions:_ The `KernelName` must be the type kernel name of a kernel that
+is defined in the application.
+The kernel `KernelName` must be compatible with the device associated with `q`
+as defined by `is_compatible`.
+
+_Returns:_ The same value `ret` that would be computed by:
+
+[source,c++]
+----
+sycl::context ctxt = q.get_context();
+sycl::device dev = q.get_device();
+auto bundle =
+  sycl::get_kernel_bundle<KernelName, sycl::bundle_state::executable>(ctxt);
+auto ret = bundle.get_kernel<KernelName>().get_info<Param>(dev);
+----
+
+_Remarks:_ Each information descriptor may specify additional preconditions,
+exceptions that are thrown, etc.
+
+
+== Issues
+
+* I purposely reduced the exceptions that are required to be thrown for certain
+  error conditions and instead listed these as preconditions.
+  An implementation can still diagnose these error conditions by throwing an
+  exception, but it is not required.
+  Since these APIs are likely on the critical path for launching a kernel, I
+  don't think we want to mandate an error check at runtime.
+  In retrospect, I think this is the right behavior for the core SYCL spec also,
+  and we should consider changing the specified behavior.
+  Thoughts?
+
+* I'm not sure how to formally specify the requirements for `KernelName`.
+  I think an implementation should be able to fail with a link-time error if
+  `KernelName` is not the type-name of some kernel that is defined in the
+  application.
+  However, this seems different from a _Constraint_, which is expected to result
+  in a compile-time error.
+  For now, I just listed it as a _Precondition_, so there is no formal
+  requirement for an implementation to diagnose this error.
```

Comment:
Seems like the specification currently uses `precondition` for template argument requirements, so I think precondition is fine.

Reply:
> Seems like the specification currently uses `precondition` for template argument requirements

In the current SYCL 2020 specification, you mean?  Yes, this is a bug in the specification, and it's my fault.  I wrote that part of the specification before I understood the difference between _Preconditions_ and _Constraints_.  Those should be _Constraints_.  It's on my list of things to clean up.